### PR TITLE
Warn on use of non-match argument to with/1 (#9764)

### PR DESCRIPTION
**Pull request by vanvoljg**

### CODE REVIEW

File: lib/elixir/src/elixir_clauses.erl [hunk 1]
```diff
@@ -401,6 +405,8 @@ format_error({try_with_only_else_clause, Origin}) ->
 
 format_error(unmatchable_else_in_with) ->
   "\"else\" clauses will never match because all patterns in \"with\" will always match";
+format_error(non_match_argument_in_with) ->
+  "\"with\" arguments should be matches, not statements";
```
Comment (josevalim) [inline lib/elixir/src/elixir_clauses.erl]:
```suggestion
  "\"with\" requires match expressions to be given either as \"left <- right\" or \"left = right\", "
  "please assign your expression to underscore if you want to simply execute some code and discard its value"
```

WDYT?

Comment (vanvoljg) [reply lib/elixir/src/elixir_clauses.erl]: Looks great, thanks!

I just realized I forgot to trap the warning output of my test. I'll use this new warning text and push a new commit.

File: lib/elixir/test/elixir/kernel/special_forms_test.exs [hunk 2]
```diff
@@ -68,4 +68,14 @@ defmodule Kernel.SpecialFormsTest do
       end
     end
   end
+
+  describe "with" do
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/special_forms_test.exs]: We have a "lib/elixir/test/elixir/kernel/warning_test.exs" file, we should most likely move it there. :)

Comment (vanvoljg) [reply lib/elixir/test/elixir/kernel/special_forms_test.exs]: I'll move it there. :-)

**Comment by vanvoljg** [comment]
I'm sorry about that last push. I'm tired and didn't watch my console output.

File: lib/elixir/src/elixir_clauses.erl [hunk 3]
```diff
@@ -401,6 +405,9 @@ format_error({try_with_only_else_clause, Origin}) ->
 
 format_error(unmatchable_else_in_with) ->
   "\"else\" clauses will never match because all patterns in \"with\" will always match";
+format_error(non_match_argument_in_with) ->
+  "\"with\" requires match expressions to be given either as \"left <- right\" or \"left = right\", "
+  "please assign your expression to underscore if you want to simply execute some code and discard its value";
```
Comment (whatyouhide) [inline lib/elixir/src/elixir_clauses.erl]:
```suggestion
  "please assign your expression to _ if you want to simply execute some code and discard its value";
```

File: lib/elixir/test/elixir/kernel/warning_test.exs [hunk 4]
```diff
@@ -1786,6 +1786,17 @@ defmodule Kernel.WarningTest do
              "extra parentheses on a remote function capture &System.pid()/0 have been deprecated. Please remove the parentheses: &System.pid/0"
   end
 
+  test "\"with\" warns when given an argument which is not a match" do
+    assert capture_err(fn ->
+             Code.eval_string("""
+                  with fn -> true end do
+                    :warn
+                  end
```
Comment (whatyouhide) [inline lib/elixir/test/elixir/kernel/warning_test.exs]: Can we indent this at the same level as the call to `Code.eval_string/1`? Not a pressing issue but would be nicer :)

File: lib/mix/lib/mix/dep/lock.ex [hunk 5]
```diff
@@ -15,7 +15,7 @@ defmodule Mix.Dep.Lock do
     opts = [file: lockfile, warn_on_unnecessary_quotes: false]
 
     with {:ok, contents} <- File.read(lockfile),
-         assert_no_merge_conflicts_in_lockfile(lockfile, contents),
+         _ <- assert_no_merge_conflicts_in_lockfile(lockfile, contents),
```
Comment (whatyouhide) [inline lib/mix/lib/mix/dep/lock.ex]:
```suggestion
         _ = assert_no_merge_conflicts_in_lockfile(lockfile, contents),
```

File: lib/elixir/src/elixir_clauses.erl (hunk 3 above)
Comment (josevalim) [reply lib/elixir/src/elixir_clauses.erl]: Let's do both?

```suggestion
  "please assign your expression to "_" (underscore) if you want to simply execute some code and discard its value";
```

**Comment by whatyouhide** [comment]
Thank you @vanvoljg! 💟

**Pull request merged**